File: Array/Merge Intervals.py
# Solution 
# Time O(n) | Space O(n)

def MergeIntervals(intervals):
    # By default it is sorted with first element
    intervals.sort()
    mergedIntervals = [intervals[0]]
    for i in range(1, len(intervals)):
        currentStart, currentEnd = mergedIntervals[-1]
        newStart, newEnd = intervals[i]
        
        if currentEnd >= newStart:
            mergedIntervals[-1][1] = max(currentEnd, newEnd)
        else:
            mergedIntervals.append(intervals[i])
            
    return mergedIntervals
  

File: array-problems/mergeIntervals.js
function mergeIntervals(intervals)
{
    // first sort the array in ascending order by first element in each interval
    intervals = intervals.sort((a,b) =>  a[0] - b[0])
    
    let previousInterval=intervals[0]
    let currentInterval

    let finalIntervals=[]
    
    for (let current=0;current < intervals.length;current++)
    {
        currentInterval=intervals[current]

        if(currentInterval[0] <= previousInterval[1])
        {
            previousInterval=[previousInterval[0],Math.max(currentInterval[1],previousInterval[1])]

        }
        else
        {
            finalIntervals.push(previousInterval)
            previousInterval=currentInterval
        }

    }

    finalIntervals.push(previousInterval)

    return finalIntervals
}

console.log(mergeIntervals([[1,3],[2,6],[8,10],[15,18]]))